caffe2/operators/batch_gather_ops.cu

```diff
@@ -37,58 +16,34 @@ bool BatchGatherOp<CUDAContext>::RunOnDevice() {
 template <>
 template <typename TInd>
 bool BatchGatherOp<CUDAContext>::DoRunWithType() {
-  auto& data = Input(DATA);
-  auto& indices = Input(INDICES);
-  auto* output = Output(0);
-
-  vector<int64_t> shape;
-  shape.push_back(data.dim(0));
-  shape.insert(shape.end(), indices.dims().begin(), indices.dims().end());
-  shape.insert(shape.end(), data.dims().begin() + 2, data.dims().end());
-  output->Resize(shape);
-
-  const int block_size = data.size_from_dim(2);
-  const int N = indices.size();
-  const auto data_batch_size = data.size_from_dim(1);
-  const auto gathered_batch_size = N * data.size_from_dim(2);
-  const TInd* idxs = indices.template data<TInd>();
-  auto src_base = static_cast<const float*>(data.raw_data());
-  auto out = static_cast<float*>(output->raw_mutable_data(data.meta()));
-  const int M = data.dim32(0);
-
-  BatchGatherKernel<<<
-      std::min(M, CAFFE_MAXIMUM_NUM_BLOCKS),
-      std::min(N * block_size, CAFFE_CUDA_NUM_THREADS),
-      0,
-      context_.cuda_stream()>>>(
-      src_base,
-      out,
-      idxs,
-      M,
-      N,
-      data_batch_size,
-      gathered_batch_size,
-      block_size);
-  return true;
+  // BatchGather is a special-case of Gather with Axis = 1, wrap = false.
+  return gather_helper::gather_impl_cuda<TInd>(
+      this, DATA, INDICES, 0, 1, false);
```

Comment:
I don't like it either but that's what I have to do if we want to share setup code and use Output(index, shape, datatype) instead of Resize. Else we'd have to copy/paste that setup code... so i think it's better to do it this was.